docs/design/task-launch-order-within-job.md

```diff
@@ -0,0 +1,113 @@
+| simple ETL | 3        | linear dependencies | no     | 1. One VCJob per phase<br />2. Manage vcjob with argo | 1.Using external solutions<br />2.Uniform management of resources across the operation is not possible<br />3.Cannot handle Running state dependencies |
+| complex ETL | n>=3     | figure dependency | Serial-parallel combination | 1. One VCJob per phase<br />2. Manage vcjob with argo| 1.Using external solutions<br />2.Uniform management of resources across the operation is not possible<br />3.Cannot handle Running state dependencies |
+| MPI        | 2        | linear dependencies | yes    | 1. Add initcontainer to the task<br/>2.Using user-written check scripts in initcontainer | 1.Increase the cost of use for users<br />2.Still using resources while waiting |
+| pytorch    | 2        | linear dependencies | yes    | 1. Add initcontainer to the task<br/>2.Using user-written check scripts in initcontainer |1.Increase the cost of use for users<br />2.Still using resources while waiting|
+| matlab     | 2        | linear dependencies | yes    | 1. Add initcontainer to the task<br/>2.Using user-written check scripts in initcontainer |1.Increase the cost of use for users<br />2.Still using resources while waiting|
+| tensorflow | n>=2   | linear dependencies | yes    | 1. Add initcontainer to the task<br/>2.Using user-written check scripts in initcontainer |1.Increase the cost of use for users<br />2.Still using resources while waiting|
+
+## Requirement
+
+Based on the scenarios listed above, the dependencies can be abstracted as:
+
+* task-A depends on task-B, which means A must be started first and then B.
+* Triggering policy, for example level triggering and edge triggering when handling circuit. In our cases, the triggering policy may be the task's event, like task running/restarted/completed. 
+
+For the ease of end user's experince, we need to unify the way of composing a complicated jobs with lots of tasks, instead of letting user handle the complexcity themselvs using init-containers or other workflow tools. So we need to  have an more advanced VCJob that has below abilities:
+
+* Each task can have one or more tasks it depends on, and it can specify the triggering policy.
+* Examples of triggering policys can be:
+  * task starts to run
+  * task restarts
+  * task completes
+* User may specify the time period after the triggering event happens, for example
+  * Imediately after task completed
+  * Half an hour afater task starts to run or completes
+* Besides the time period, user may specify an certain condition as the triggering policy, for example,
+  * Check if a specified file exists in the /data directory.
+  * Check if the specified port is open, such as 2222.
+  * Check if the HTTP GET request to the specified path was successfully accessed, such as /health.
+  
+## Design
+
+### API
+taskflow example
+```
+apiVersion: batch.volcano.sh/v1alpha1
+kind: TaskFlow
+metadata:
+  name: task-flow
```

Comment:
task or job?

Reply:
if job, how to handle MPI job issue?

Reply:
I think it is `taskflow`.
if mpi, we can create a taskflow like this:

```yaml
apiVersion: batch.volcano.sh/v1alpha1
kind: TaskFlow
metadata:
  name: task-flow
spec:
  tasks:
  - name: worker
  - name: master
    depends: 
      target: ['worker']
      probe:
        taskStatus: Running

```

Reply:
if so, why not just add `Depends` to volcano job?

Reply:
We think this taskflow can be reused by other vcjob.
@Thor-wl Can you join this discussion?

Reply:
> We think this taskflow can be reused by other vcjob.

Why reuse? If Job_1 has task A, B, C and Job_2 has task B, C, D, how can you sure those two job works as expected?

Reply:
If we want to reuse a taskflow, then we need to satisfy the following two conditions:

* vcjob's taskflow field value is the name of the taskflow.
* The task in vcjob has the same name as the task in taskflow

for example:
```
jobA: 
task: taskA, taskB, taskC
taskflow: taskflowA

jobB: 
task: taskA, taskB,taskC
taskflow: taskflowA

taskflowA: 
task: taskA, taskB, taskC
```
In this case, both jobA and jobB can use taskflowA